build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/GoRun.java

```diff
@@ -0,0 +1,220 @@
+    private static final String LOCAL_GO_SUM_BACKUP_FILENAME = "jfrog.go.sum.backup";
+    private static final String LOCAL_GO_MOD_BACKUP_FILENAME = "jfrog.go.mod.backup";
+
+    private List<Dependency> dependenciesList = Lists.newArrayList();
+    private CommandExecutor goCommandExecutor;
+    private String goCmdArgs;
+    private String resolutionRepository;
+    private String resolverUsername;
+    private String resolverPassword;
+    private Map<String, String> env;
+
+    /**
+     * Run go command and collect dependencies.
+     *
+     * @param goCmdArgs     - Go cmd args.
+     * @param path          - Path to directory contains go.mod.
+     * @param clientBuilder - Client builder for resolution.
+     * @param repo          - Artifactory's repository for resolution.
+     * @param username      - Artifactory's username for resolution.
+     * @param password      - Artifactory's password for resolution.
+     * @param logger        - The logger.
+     * @param env           - Environment variables to use during npm execution.
+     */
+    public GoRun(String goCmdArgs, Path path, ArtifactoryBuildInfoClientBuilder clientBuilder, String repo, String username, String password, Log logger, Map<String, String> env) throws IOException {
+        super(clientBuilder, path, logger);
+        this.goCmdArgs = goCmdArgs;
+        this.resolutionRepository = repo;
+        this.resolverUsername = username;
+        this.resolverPassword = password;
+        this.env = env;
+    }
+
+    public Build execute() {
+        try (ArtifactoryBuildInfoClient dependenciesClient = (clientBuilder != null ? (ArtifactoryBuildInfoClient) clientBuilder.build() : null)) {
+            if (dependenciesClient != null) {
+                client = dependenciesClient;
+                preparePrerequisites(resolutionRepository, client);
+                setResolverAsGoProxy();
+            }
+            this.goCommandExecutor = new CommandExecutor(GO_CLIENT_CMD, env);
+            runCmd();
+            collectDependencies();
+            return createBuild(null, dependenciesList);
+        } catch (Exception e) {
+            logger.error(e.getMessage(), e);
+        }
+        return null;
+    }
+
+    /*
+     * In order to use Artifactory as a resolver we need to set GOPROXY env. var with Artifactory details.
+     * Wa also support fallback to VCS in case pkg doesn't exist in Artifactort,
+     */
+    private void setResolverAsGoProxy() throws IOException, URISyntaxException {
+        URL rtUrl = new URL(client.getArtifactoryUrl());
+        URIBuilder proxyUrlBuilder = new URIBuilder()
+                .setScheme(rtUrl.getProtocol())
+                .setUserInfo(resolverUsername, resolverPassword)
+                .setHost(rtUrl.getHost())
+                .setPath(rtUrl.getPath() + ARTIFACTORY_GO_API + resolutionRepository);
+        String proxyValue = proxyUrlBuilder.build().toURL().toString() + "," + GOPROXY_VCS_FALLBACK;
```

Comment:
Can this approach also be implemented in JFrog CLI? I think it is implemented differently there.

Reply:
Do you mean using "direct" in GOPROXY value?
If yes - the fallback approach is implemented differently in gocmd:
 First, we try to get the pkg from Artifactory and if we get an error we try to get it from the VCS server.
 I think we can change it and let the Go client do the fallback by itself.